### PR TITLE
direct gauges: Support unaggregated gauges

### PR DESCRIPTION
**Pull request by theatrus**
Replaces the previously scary and mostly useless KV type with a
direct_gauge type (type 'G' or 'k') which only reports the last set
value in all cases.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -276,16 +276,12 @@ Currently supported message types:
 After the flush interval, the counters and timers of the same key are
 aggregated and this is sent to the store.
 
-Gauges also support "delta" updates, which are supported by prefixing the
-value with either a `+` or a `-`. This implies you can't explicitly set a gauge to a negative number without first setting it to zero.
+Aggregaed gauges also support "delta" updates, which are supported by prefixing
```
Comment (golfdish) [inline README.md:279]: Aggrega`t`ed

File: src/gauge_direct.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * @arg gauge The gauge struct to initialize
+ * @return 0 on success.
+ */
+int init_gauge_direct(gauge_direct_t *gauge);
+
+/**
+ * Adds a new sample to the struct
+ * @arg gauge The gauge to add to
+ * @arg sample The new sample value
+ * @arg delta   Is this a delta update?
```
Comment (golfdish) [inline src/gauge_direct.h:22]: answer: no

File: src/metrics.h [hunk 3]
```diff
@@ -24,15 +19,15 @@ typedef struct {
...
-    hashmap *gauges;    // Map of name -> guage struct
-    key_val *kv_vals;   // Linked list of key_val structs
-    double timer_eps;   // The error for timers
-    double *quantiles;  // Array of quantiles
-    uint32_t num_quants; // Size of quantiles array
-    radix_tree *histograms; // Radix tree with histogram configs
+    hashmap *counters;           // Hashmap of name -> counter structs
+    hashmap *timers;             // Map of name -> timer_hist structs
+    hashmap *sets;               // Map of name -> set_t structs
+    hashmap *gauges;             // Map of name -> guage struct
```
Comment (golfdish) [inline src/metrics.h:25]: `guage` 🤔

File: README.md (hunk 1 above)
Comment (theatrus) [reply README.md]: t aplied

File: src/gauge_direct.h (hunk 2 above)
Comment (theatrus) [reply src/gauge_direct.h]: correct

File: src/metrics.h (hunk 3 above)
Comment (theatrus) [reply src/metrics.h]: OG typos!

**Comment by maghamravi** [comment]
@theatrus Thanks for coming up with direct gauges! This will immensely help Flink / Beam pipelines.  Curious, any timeline for us to adopt and test this ?  cc @badiib

**Comment by theatrus** [comment]
@maghamravi We'll do a rollout of the needed components this week. Will ping you internally when this is out.

**Pull request merged**